#162948841 create profiles

kwameasiago opened the pull request:
**What does this PR do?**
Create an endpoint for users to edit profiles

**Description of tasks to be completed**
- Created an endpoint to update username, first name, last name, bio, profile image
- Created tests for editing profile

**How should this be manually tested?**
- `git clone https://github.com/andela/ah-django-unchained`
- `cd ah-django-unchained`
- `git checkkout ft-create-profile-162948841`
- `python3 -m venv virtual_env_name` to create a virtual environment
- Add a `.env` file as indicated in the ReadMe
- Add the following to your env

   export CLOUD_NAME='authors'
   export API_KEY='777468436252743'
   export API_SECRET='_mRwkwGpbRKurTYgzSuSPlxG3uk'

- `source .env` to load your environment variables
- `pip install -r requirements.txt`
- Run the migrations `python manage.py makemigrations && python manage.py migrate`

**What are the relevant Pivotal Tracker Stories?**
[#162948841](https://www.pivotaltracker.com/story/show/162948841)

**Screenshots**
![screenshot 2019-01-19 at 13 20 39](https://user-images.githubusercontent.com/30651751/51425426-17159400-1bed-11e9-99cc-a8475def339b.png)

kwameasiago committed:
Feature (Profiles) Add profile endpoints
Create profile app
Add first_name, last_name, bio and profile image models
create profile serializer
create profile views
[Delivers #162948841]

kwameasiago committed:
feat (Profile) user can create a profile
- Add post save and pre save signals
- Add a slug
- Clean up code

[Delivers #162948841]

kwameasiago committed:
feat (Profile) user can create a profile
- Create a test folder
[Delivers #162948841]

kwameasiago committed:
feat(profile test)Add test

- Add post profile test
- Add get profile test

[Delivers #162948841]

andrewhingah commented:
Hey Kwame. Please include some screenshots

andrewhingah commented on authors/apps/authentication/admin.py:
```diff
@@ -0,0 +1,5 @@
+
+from .models import User
+
+admin.site.register(User)
```
Add an extra line after this

MaggieChege replied:
Add a blankline

Georgeygigz replied:
add a blank line

MaggieChege commented on authors/settings.py:
```diff
@@ -170,4 +170,9 @@
     },
 }
+cloudinary.config(
+     cloud_name = os.getenv("CLOUD_NAME"),
```
Add .env configuratations in your PR and also add them to the .env.sample

kwameasiago committed:
feat (feedback) Implement feedback
- Add new line at the end of the file

[Delivers #162948841]

MaggieChege commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,40 @@
+            "first_name":"","last_name":"",
+            "gender":"M","bio":"","profile_image":None}
+
+    def test_get_profile(self):
```
Make you code easier to real. Adhere t PEP8 style Guide

kwameasiago commented:
> Hey Kwame. Please include some screenshots

I have

andrewhingah commented:
Should I use the same API secret keys? If not, please advise on how I can put them in a `.env` rather than exposing them.

Georgeygigz commented on authors/apps/core/permissions.py:
```diff
@@ -29,3 +29,20 @@ def has_object_permission(self, request, view, obj):
+            return True
+
+        # write permission is only allowed to the owner of the object
+        print(obj)
```
Kindly remove print

kwameasiago commented:
> Should I use the same API secret keys? If not, please advise on how I can put them in a `.env` rather than exposing them.

Yes, that is the API key to cloudinary. Please add it to your env

kwameasiago committed:
feat (Profile) Implement feedback

- remove the print

[Delivers #162948841]

kwameasiago committed:
feat (Profile) Implement feedback

- Make function easier to read

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/core/permissions.py:
```diff
@@ -29,3 +29,19 @@ def has_object_permission(self, request, view, obj):
         return obj.author != request.user
+
+
+class IsMyProfileOrReadOnly(BasePermission):
```
The class IsAuthorOrReadOnly was created for this purpose, use that instead.

kwameasiago replied:
Because my profile model is referencing the user model via a one-to-one relationship I had to override **obj.author != request.user**  in the IsAuthorOrReadOnly class with 
**obj.user.username == request.user.username** in the IsMyProfileOrReadOnly class

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+from django.db import models
+from django.db.models.signals import post_save, pre_save
+from django.utils.text import slugify
+from ..authentication.models import User,UserManager
```
Third party packages before ours

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+
+    # A choice field to hold multiple choice 
+    gender_choices = (('M','MALE'),('F','FEMALE'))
+    gender = models.CharField(max_length=10, choices=gender_choices, default='M')
```
Why not make this nullable field instead of having Male as default?

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+    gender = models.CharField(max_length=10, choices=gender_choices, default='M')
+
+    # cloudinary field
+    profile_image = CloudinaryField('image', null=True, default='', blank=True)
```
No need to have default if it's empty

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+
+    # A char field for the user Last name
+    last_name = models.CharField(max_length=100, default='')
+    objects = UserManager()
```
We should have this after all the model fields so it's not confusing

ignatiusukwuoma commented on authors/apps/profiles/urls.py:
```diff
@@ -0,0 +1,9 @@
+
+app_name = "profiles"
+urlpatterns = [
+    path('profile/<slug:slug>/', ProfileView.as_view(), name='post-profile'),
```
paths should be `profiles/` not `profile/` and the slug should be username

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+    objects = UserManager()
+
+    # create a slug for the user name
+    slug = models.SlugField(null=True, blank=True)
```
What's the purpose of the slug? Why don't we just use the username as recommended in the API Spec in the Readme

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,52 @@
+import json
```
This import is not used

kwameasiago replied:
Fixed it

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,52 @@
+        profile = self.client.get(reverse('profiles:get-profile', kwargs={'slug': 'johndoe'}),format='json')
+        self.assertEqual(self.profile_data,profile.data)
+    
+    def test_post_profile(self):
```
This is testing update profile, test name doesn't accurately reflect test job

kwameasiago replied:
I have fixed it

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,52 @@
+        # override self.profile_data 
+        self.profile_data['first_name'],self.profile_data['last_name'] = 'kwame', 'asiago'
+        self.profile_data['bio'] = 'some bio'
+        del self.profile_data['profile_image']
```
This test looks complicated, please simplify it. Consider, creating new variables like the profile_data you need instead of overriding what's already available.

kwameasiago replied:
I have simplified the test

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,52 @@
+        profile = self.client.put(reverse('profiles:post-profile', kwargs={'slug': 'johndoe'}),self.profile_data,
+        format='json',HTTP_AUTHORIZATION='token {}'.format(token))
+        self.profile_data['profile_image'] = None
+        self.assertEqual(self.profile_data,profile.data)
```
Add more tests

kwameasiago replied:
I have added more test

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,22 @@
+    permission_classes = (IsMyProfileOrReadOnly,)
+    lookup_field = 'slug'
+
+class GetProfileView(generics.ListAPIView):
```
The names of the views look very much alike, they should be more different

ignatiusukwuoma commented:
Run your files though pycodestyle and ensure your codes follows PEP8

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,50 @@
+    first_name = models.CharField(max_length=100, default='')
+
+    # A char field for the user Last name
+    last_name = models.CharField(max_length=100, default='')
```
Let's have a timestamp field, probably `updated_at` so we keep record of when profile was edited. We can always get `created_at` from User.

kwameasiago committed:
feat(profiles) Implement feedback
- Edit to pep8 standard
- remove json import from test_user_profile.py
- reorder importations
[Delivers #162948841]

kwameasiago committed:
feat(profiles) Implement feedback
- Make profile test more readable
[Delivers #162948841]

kwameasiago committed:
feat(profiles) Implement feedback
- change url name to profiles
[Delivers #162948841]

kwameasiago committed:
feat(profile) Implement feedback

 - Add tests for empty values and invalid gender
- Format to pep8 standards
- Add extra field N to gender

[Delivers #162948841]

Georgeygigz commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+            user_instance = UserProfile.objects.select_related(
+                'user').get(user__username=username)
+        except:
+            return Response({'response': 'Username {} not found'.format(username)}, status.HTTP_404_NOT_FOUND)
```
Line too long, consider using pep8 standards

Georgeygigz commented on authors/apps/profiles/serializers.py:
```diff
@@ -0,0 +1,35 @@
+        write_only=True,
+        error_messages={
+            'required': 'Password field required',
+            'invalid': 'Please enter M if you are male, F if you are female or N if you do not want to disclose '
```
Line too long consider using pep8 to format your code.

kwameasiago committed:
feat(profile) Implement feedback

- shorten my test length

[Delivers #162948841]

MaggieChege commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,93 @@
+    def test_get_user_profile(self):
+        """Test get profile upon registrations"""
+        # register user
+        register = self.client.post(
```
Avoid repetition in your tests.

kwameasiago replied:
Fixed it

ignatiusukwuoma replied:
You can use the function I recommended above to register `johndoe`.

andrewhingah commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,93 @@
+        profile = self.client.put(reverse('profiles:put-profile', kwargs={'username': 'johndoe'}),
+                                  self.invalid_gender, format='json', HTTP_AUTHORIZATION='token {}'.format(token))
+        self.assertEqual(profile.status_code, status.HTTP_400_BAD_REQUEST)
+        data = {'errors': {'gender': ['Please enter M if you are male, F if you are female or '
```
Just define a message here i.e. `msg = 'Please enter M if you are male, F if you are female or ....` and then compare it with the response, instead of having all the dictionaries and a list.

kwameasiago replied:
fixed it

kwameasiago committed:
feat(profile) Implement feedback

- Create a function for getting token

[Delivers #162948841]

kwameasiago committed:
feat(profile) Implement feedback

- Create a variable for the variable

[Delivers #162948841]

MaggieChege reviewed (approved):
Looks good to me

ignatiusukwuoma commented:
You will need to rebase your branch and add tests if necessary so coverage doesn't decrease

kwameasiago committed:
feat (profile) Update file
- Update requirement.txt
[Delivers #162948841]

kwameasiago committed:
feat (profile) Fix permission class

- Fix update permission class

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,58 @@
+    """
+    This class contains  models relating to user profile. 
+    It has a one to one relationship with User model class.
+    An intance of this class is created during user registration by calling 
```
typo - instance

kwameasiago replied:
Fixed it

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,58 @@
+
+    user = models.OneToOneField(User, on_delete=models.CASCADE)
+    # A char field to hold bio data
+    bio = models.TextField(max_length=200, default='empty')
```
We don't need a `default`, do we? And this should be a blank=True fielld

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,58 @@
+    # A choice field to hold multiple choice
+    gender_choices = (('M', 'MALE'), ('F', 'FEMALE'),
+                      ('N', 'I PREFER NOT TO SAY'))
+    gender = models.CharField(
```
Why not leave as blank until I select a gender

kwameasiago replied:
I thought it would be user-friendly to ask for clarification for the user. So the user knows they don't have to provide a gender

ignatiusukwuoma commented on authors/apps/profiles/models.py:
```diff
@@ -0,0 +1,58 @@
+    instance.slug = slug
+
+
+pre_save.connect(create_slug_receiver, sender=UserProfile)
```
Are we still making use of slugs?

kwameasiago replied:
No. Apologies, I forgot to remove that function

kwameasiago committed:
feat(profile) Implement feedback

- Add more test
- Edit test comments

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/profiles/serializers.py:
```diff
@@ -0,0 +1,35 @@
+        regex='^(N|F|M)$',
+        write_only=True,
+        error_messages={
+            'required': 'Password field required',
```
What is Password field doing in genders?

kwameasiago replied:
FIxed it

kwameasiago committed:
feat(profile) Implement feedback

- Remove slug

[Delivers #162948841]

kwameasiago committed:
feat(profile) Implement feedback

- Remove password field in gender

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/profiles/serializers.py:
```diff
@@ -0,0 +1,35 @@
+        gender = data.get('gender', None)
+        if first_name is None:
+            raise serializers.ValidationError('first name was not provided')
+        if first_name is None:
```
Shouldn't this be last_name? But come to think of it, do we really need to overwrite the validate function? Our model fields should handle this already. However, for someone whose Profile was created automatically after registration as a user, should he not be allowed to fill in any of the data when he wants?

kwameasiago replied:
I have removed the validator for first name and last name

ignatiusukwuoma commented on authors/apps/profiles/urls.py:
```diff
@@ -0,0 +1,8 @@
+
+app_name = "profiles"
+urlpatterns = [
+    path('profiles/<str:username>/', ProfileView.as_view(), name='put-profile'),
```
What's wrong with using <username> without the str?

kwameasiago replied:
Please clarify.

ignatiusukwuoma replied:
WHy not just `<username>` instead of `<str:username>`

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+        try:
+            user_instance = UserProfile.objects.select_related(
+                'user').get(user__username=username)
+            if user_instance.user.username != request.user.username:
```
I thought you had a permission class that was handling this?

kwameasiago replied:
I did but when I override the update feature the I had to check the permission manually

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+                'user').get(user__username=username)
+            if user_instance.user.username != request.user.username:
+                return Response({'error':'You are not allowed to edit or delete this object'})
+        except:
```
except what? The exception is too broad

kwameasiago replied:
Fixed it by adding a does not exist exception

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.update(instance=user_instance, validated_data=request.data)
+        return Response({'response': 'profile has been updated successfully '})
```
Let's stick to using 'message', instead of 'response'

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+        try:
+            user_instance = UserProfile.objects.select_related(
+                'user').get(user__username=username)
+        except:
```
ditto

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+            user_instance = UserProfile.objects.select_related(
+                'user').get(user__username=username)
+        except:
+            return Response({'': ''}, status.HTTP_404_NOT_FOUND)
```
What's the response?

kwameasiago replied:
Fixed it

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,34 @@
+        except:
+            return Response({'': ''}, status.HTTP_404_NOT_FOUND)
+        serializer = self.serializer_class(user_instance)
+        return Response({'profile': serializer.data})
```
Wouldn't you like to specify the status code?

kwameasiago replied:
Fixed it

ignatiusukwuoma commented on authors/urls.py:
```diff
@@ -27,6 +27,7 @@
     path('api/', include('authors.apps.authentication.urls',
                          namespace='authentication')),
     path('api/', include('authors.apps.friends.urls')),
+    path('user/', include('authors.apps.profiles.urls', namespace='profiles')),
```
'user/' here should be 'api/'

kwameasiago replied:
Fixed it

kwameasiago committed:
feat(profile) Implement feedback

- Fix typo in userprofile serializers

[Delivers #162948841]

kwameasiago committed:
feat(profile) Implement feedback

- Add status code
- Change status from response to message
- Add exceptions does not exist

[Delivers #163346018]

kwameasiago committed:
feat(profile) Implement feedback

- Remove first name and last name validators

[Delivers #163346018]

kwameasiago committed:
feat(profile) Implement feedback

- Remove default in bio model

[Delivers #163346018]

kwameasiago committed:
feat(profile) Implement feedback

- Format to pep8

[Delivers #163346018]

kwameasiago committed:
feat(profile) Implement feedback

- Add comments to models

[Delivers #163346018]

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,35 @@
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.update(instance=user_instance, validated_data=request.data)
+        return Response({'message': 'profile has been updated successfully '},status.HTTP_201_CREATED)
```
Response status of 200 will be more appropriate as nothing new was created. Also, start your sentence with capital letters and remove the space at the end of the response sentence.

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+        self.assertEqual(register.status_code, status.HTTP_201_CREATED)
+        return register.data['token']
+
+    def get_token_jane(self):
```
This is the same with the function above. Modify the `get_token` function to take an argument(signup details) and return the token based on the details provided. We should rename the function to something like `register_user` Then you don't need this function.

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+
+        # get user profile
+        profile = self.client.get(
+            reverse('profiles:put-profile', kwargs={'username': 'johndoe'}), format='json')
```
On a first view I won't understand where 'johndoe' is coming from, but if you use `self.data_john.username` it is clear that it's the username declared above and I don't have to scroll up to check.

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+        """Test updating profile upon registrations"""
+        token = self.get_token()
+        # update profile
+        profile = self.client.put(reverse('profiles:put-profile', kwargs={'username': 'johndoe'}),
```
Nothing wrong.. but you're updating with same username

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+            self.invalid_gender,
+            format='json',
+            HTTP_AUTHORIZATION='token {}'.format(token))
+        # self.assertEqual(profile.status_code, status.HTTP_400_BAD_REQUEST)
```
Remove commented code if not useful

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+        self.assertEqual(data, json.loads(profile.content))
+
+    def test_updating_invalid_user(self):
+        """Test updating with other peaople user porofile"""
```
typos

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,117 @@
+            reverse(
+                'profiles:put-profile',
+                kwargs={'username': 'janedoe'}),
+            self.invalid_gender,
```
Please make this test focus on one thing, is it updating with invalid gender or is it trying to update another person's profile?

ignatiusukwuoma commented:
Please stop creating commit messages like `implement feedback`. Write what was done in that commit... the feedback that was implemented.

kwameasiago committed:
feat(profile) Edit test_user_profiile.py
- Remove duplicated method get_token
- Change get_token to register_user
- Remove typo
- Replace 'johndoe' with self.data_john['user']['username']
- Formated to Pep8

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,106 @@
+        self.assertEqual(profile.status_code, status.HTTP_400_BAD_REQUEST)
+        msg = 'Please enter M if you are male, F if you are female or N if you do not want to disclose'
+        data = {'errors': {'gender': [msg]}}
+        self.assertEqual(data, json.loads(profile.content))
```
Just so you know... `profile.content` is A bytestring which you have to json.loads on... but profile.data is a python object that will equal your `data` variable without json.loads

kwameasiago replied:
I had tried with data but it was not returning with the correct data

ignatiusukwuoma commented on authors/apps/profiles/urls.py:
```diff
@@ -0,0 +1,8 @@
+
+
+app_name = "profiles"
+urlpatterns = [
```
WHy not just `<username>` instead of `<str:username>`

ignatiusukwuoma commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,35 @@
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.update(instance=user_instance, validated_data=request.data)
+        return Response({'message': 'Profile has been updated successfully'},status.HTTP_200_OK)
```
Please return the user's profile after an update, so that on the frontend we won't need need to make another get request to get the user's new details

kwameasiago committed:
feat(profile) Change return statement

- Return the data updated after an update

[Delivers #162948841]

kwameasiago committed:
feat(profile) Add keyword profile

- Return with keyword profile as stated in readme

[Delivers #162948841]

kwameasiago committed:
feat(profile) Change profile url

- Change <str: username> to <username>

[Delivers #162948841]

kwameasiago committed:
feat(profile) Fix failing test

- Change assert statement

[Delivers #162948841]

KMaina commented on authors/apps/profiles/views.py:
```diff
@@ -0,0 +1,36 @@
+            user_instance = UserProfile.objects.select_related(
+                'user').get(user__username=username)
+            if user_instance.user.username != request.user.username:
+                return Response({'error': 'You are not allowed to edit or delete this object'}, status.HTTP_403_FORBIDDEN)
```
Try having `{'error': 'You are not allowed to edit or delete this object'}` as a variable to make to easier to read. Do this across your app.

kwameasiago committed:
feat(profile) Make view readable

- Create a response variable for update view

[Delivers #162948841]

ignatiusukwuoma commented on authors/apps/profiles/test/test_user_profile.py:
```diff
@@ -0,0 +1,105 @@
+            self.invalid_gender,
+            format='json',
+            HTTP_AUTHORIZATION='token {}'.format(token))
+        # self.assertEqual(profile.status_code, status.HTTP_400_BAD_REQUEST)
```
Commented test?
